streams/src/main/java/org/apache/kafka/streams/state/internals/StoreChangeLogger.java

```diff
@@ -109,7 +109,7 @@ public void logChange(ValueGetter<K, V> getter) {
             }
             for (K k : this.dirty) {
                 V v = getter.get(k);
-                collector.send(new ProducerRecord<>(this.topic, this.partition, k, v), keySerializer, valueSerializer);
+                collector.send(new ProducerRecord<>(this.topic, this.partition, context.timestamp(), k, v), keySerializer, valueSerializer);
```

Comment:
I think this may help with log rolling, but not with log retention (I should be more specific in my previous comment). More specifically, in `TimestampExtractor` we describe the extracted timestamp as:

```
/**
 * An interface that allows the Kafka Streams framework to extract a timestamp from an instance of {@link ConsumerRecord}.
 * The extracted timestamp is defined as milliseconds.
 */
public interface TimestampExtractor {

    /**
     * Extracts a timestamp from a record.
     * <p>
     * Typically, the timestamp represents the milliseconds since midnight, January 1, 1970 UTC.
     *
     * @param record  a data record
     * @return        the timestamp of the record
     */
    long extract(ConsumerRecord<Object, Object> record);
}
```

but we do not enforce the extracted timestamp must to be `unix timestamp since epoch`, that indicating `when the record was created`. I.e. users can extract timestamps with any definitions. On the other hand, timestamp based log rolling / log retention assumes the message timestamp to be `unix timestamp since epoch`, and are comparing that timestamp with its own system time to determine when to roll / truncate a log segment.

I think this problem is general in Kafka Streams, since without restricting the timestamp semantics our definition on the windowing operators are actually

```
window with XXX time units, that needs to be retained for YYY time units
```

where `time units` are not necessarily `milliseconds since epoch`. And similarly with the punctuate function it is:

```
punctuate every ZZZ time units
```

I'm thinking we can probably either comment in the doc string that `it MUST be representing the milliseconds since midnight, January 1, 1970 UTC... And note that it is used for log retention blah blah, so if you don't set it right, your fault.` Or not using the event timestamp for changelog records (for other issues like punctuate we can discuss separately), but rely on system time only, where the definition becomes:

```
window with XXX time units, that needs to be retained for YYY milliseconds in system time.
```

Admittedly this is a bit awkward regarding the terminology. Thoughts?